src/http/http_client.f90

```diff
@@ -32,42 +32,39 @@ module http_client
     function new_request(url, method, header) result(response)
         character(len=*), intent(in) :: url
         integer, intent(in), optional :: method
-        type(header_type), intent(inout), optional :: header
-        type(header_type) :: default_header
+        type(header_type), allocatable, intent(inout), optional :: header(:)
+        type(header_type), allocatable :: default_header(:)
         type(request_type) :: request
         type(response_type) :: response
         type(client_type) :: client
 
-        ! setting default HTTP method
-        if(present(method)) then 
-           request%method = method
-        else
-            request%method = 1
-        end if
+        ! Set default HTTP method.
+        request%method = optval(method, 1)
         
-        ! setting defautl request headers
+        ! Set default request headers.
         if(present(header)) then 
             call set_default_headers(header)
             request%header = header
         else
             call set_default_headers(default_header)
             request%header = default_header
         end if
-        ! populate the request header_key array
-        call request%header%set_header_key()
+
         ! setting request url
         request%url = url
+        
         client = client_type(request=request)
-        response = client%client_get_response()
         
+        ! Populates the response 
+        response = client%client_get_response()
     end function new_request
+
     ! Constructor for client_type type.
     function new_client(request) result(client)
         type(request_type), intent(in) :: request
         type(client_type) :: client
 
         client%request = request
-
     end function new_client
```

Comment:
Now you could even get rid of this constructor and use the default type constructor i.e. `client_type(request)` to instantiate it. But then, since all `client_type` does is pass the `request` instance to return the `response` instance, we don't even need the `client_type` at all.

It would be worthwhile keeping `client_type` if you anticipate needing to maintain some state that doesn't belong in either `request_type` or `response_type`.